File: src/main/java/me/zeroeightsix/kami/util/Friends.java
package me.zeroeightsix.kami.util;

import com.google.common.base.Converter;
import com.google.gson.JsonElement;
import com.google.gson.JsonParser;
import com.google.gson.JsonPrimitive;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.URL;
import java.net.URLConnection;
import java.util.ArrayList;
import java.util.UUID;
import java.util.regex.Pattern;

/**
 * Created by 086 on 13/12/2017.
 */
public class Friends {
    public static final Friends INSTANCE = new Friends();

    public static Setting<ArrayList<Friend>> friends;

    private Friends() {
    }

    public static void initFriends() {
        friends = Settings.custom("Friends", new ArrayList<Friend>(), new FriendListConverter()).buildAndRegister("friends");
    }

    public static boolean isFriend(String name) {
        return friends.getValue().stream().anyMatch(friend -> friend.username.equalsIgnoreCase(name));
    }

    public static class Friend {
        String username;
        UUID uuid;

        public Friend(String username, UUID uuid) {
            this.username = username;
            this.uuid = uuid;
        }

        public String getUsername() {
            return username;
        }
    }

    public static class FriendListConverter extends Converter<ArrayList<Friend>, JsonElement> {
        public FriendListConverter() {
        }

        @Override
        protected JsonElement doForward(ArrayList<Friend> list) {
            StringBuilder present = new StringBuilder();
            for (Friend friend : list)
                present.append(String.format("%s;%s$", friend.username, friend.uuid.toString()));
            return new JsonPrimitive(present.toString());
        }

        @Override
        protected ArrayList<Friend> doBackward(JsonElement jsonElement) {
            String v = jsonElement.getAsString();
            String[] pairs = v.split(Pattern.quote("$"));
            ArrayList<Friend> friends = new ArrayList<>();
            for (String pair : pairs) {
                try {
                    String[] split = pair.split(";");
                    String username = split[0];
                    UUID uuid = UUID.fromString(split[1]);
                    friends.add(new Friend(getUsernameByUUID(uuid, username), uuid));
                } catch (Exception ignored) {
                } // Empty line, wrong formatting or something, we don't care
            }
            return friends;
        }

        private String getUsernameByUUID(UUID uuid, String saved) {
            String src = getSource("https://sessionserver.mojang.com/session/minecraft/profile/" + uuid.toString());
            if (src == null || src.isEmpty()) return saved;
            try {
                JsonElement object = new JsonParser().parse(src);
                return object.getAsJsonObject().get("name").getAsString();
            } catch (Exception e) {
                e.printStackTrace();
                System.err.println(src);
                return saved;
            }
        }

        private static String getSource(String link) {
            try {
                URL u = new URL(link);
                URLConnection con = u.openConnection();
                BufferedReader in = new BufferedReader(new InputStreamReader(con.getInputStream()));
                StringBuilder buffer = new StringBuilder();
                String inputLine;
                while ((inputLine = in.readLine()) != null)
                    buffer.append(inputLine);
                in.close();

                return buffer.toString();
            } catch (Exception e) {
                return null;
            }
        }
    }

}

File: src/main/java/me/zeroeightsix/kami/util/Friends.kt
package me.zeroeightsix.kami.util

import com.google.gson.GsonBuilder
import com.google.gson.JsonParser
import com.google.gson.annotations.SerializedName
import com.google.gson.reflect.TypeToken
import com.mojang.util.UUIDTypeAdapter
import me.zeroeightsix.kami.KamiMod
import me.zeroeightsix.kami.manager.mangers.FileInstanceManager
import me.zeroeightsix.kami.manager.mangers.FriendManager
import me.zeroeightsix.kami.util.text.MessageSendHelper
import java.io.*
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets
import java.util.*
import kotlin.collections.ArrayList

/**
 * Created by 086 on 13/12/2017.
 * Updated by Xiaro on 14/08/20
 */
object Friends {

    @JvmStatic
    fun isFriend(name: String): Boolean {
        return FriendManager.friendFile.enabled && FriendManager.friendFile.friends.any { friend -> friend.username.equals(name, ignoreCase = true) }
    }

    fun addFriend(name: String): Boolean {
        val friend = getFriendByName(name)
        return if (friend == null) {
            false
        } else {
            FriendManager.friendFile.friends.add(friend)
            true
        }
    }

    fun removeFriend(name: String): Boolean {
        return FriendManager.friendFile.friends.removeIf { friend -> friend.username.equals(name, ignoreCase = true) }
    }

    fun getFriendByName(input: String): Friend? {
        val infoMap = ArrayList(Wrapper.minecraft.connection!!.playerInfoMap)
        val profile = infoMap.find { info ->
            info!!.gameProfile.name.equals(input, ignoreCase = true)
        }

        return if (profile != null) {
            Friend(profile.gameProfile.name, profile.gameProfile.id)
        } else {
            MessageSendHelper.sendChatMessage("Player isn't online. Looking up UUID..")
            val s = requestIDs(input)
            if (s.isNullOrBlank()) {
                MessageSendHelper.sendChatMessage("Couldn't find player ID. Are you connected to the internet? (0)")
                null
            } else {
                val element = JsonParser().parse(s)
                if (element.asJsonArray.size() == 0) {
                    MessageSendHelper.sendChatMessage("Couldn't find player ID. (1)")
                    null
                } else {
                    try {
                        val id = element.asJsonArray[0].asJsonObject["id"].asString
                        val username = element.asJsonArray[0].asJsonObject["name"].asString
                        Friend(username, UUIDTypeAdapter.fromString(id))
                    } catch (e: Exception) {
                        e.printStackTrace()
                        MessageSendHelper.sendChatMessage("Couldn't find player ID. (2)")
                        null
                    }
                }
            }
        }
    }

    private fun requestIDs(input: String): String? {
        val data = "[\"$input\"]"
        return try {
            val url = URL("https://api.mojang.com/profiles/minecraft")
            val conn = url.openConnection() as HttpURLConnection
            conn.connectTimeout = 5000
            conn.setRequestProperty("Content-Type", "application/json; charset=UTF-8")
            conn.doOutput = true
            conn.doInput = true
            conn.requestMethod = "POST"
            val os = conn.outputStream
            os.write(data.toByteArray(StandardCharsets.UTF_8))
            os.close()

            // read the response
            val `in`: InputStream = BufferedInputStream(conn.inputStream)
            val res = convertStreamToString(`in`)
            `in`.close()
            conn.disconnect()
            res
        } catch (e: Exception) {
            null
        }
    }

    private fun convertStreamToString(`is`: InputStream): String {
        val s = Scanner(`is`).useDelimiter("\\A")
        return if (s.hasNext()) s.next() else "/"
    }

    class Friend(
            @SerializedName("Name")
            var username: String,

            @SerializedName("UUID")
            var uuid: UUID
    )
}